chips/cc26xx/src/gpio.rs

```diff
@@ -0,0 +1,190 @@
+        }
+    }
+
+    fn enable_gpio(&self) {
+        ioc::IOCFG[self.pin].enable_gpio();
+    }
+
+    pub fn set_client<C: hil::gpio::Client>(&self, client: &'static C) {
+        self.client.set(Some(client));
+    }
+
+    pub fn handle_interrupt(&self) {
+        self.client.get().map(|client| {
+            client.fired(self.client_data.get());
+        });
+    }
+}
+
+impl hil::gpio::PinCtl for GPIOPin {
+    fn set_input_mode(&self, mode: hil::gpio::InputMode) {
+        ioc::IOCFG[self.pin].set_input_mode(mode);
+    }
+}
+
+impl hil::gpio::Pin for GPIOPin {
+    fn make_output(&self) {
+        self.enable_gpio();
+        // Disable input in the io configuration
+        ioc::IOCFG[self.pin].enable_output();
+        // Enable data output
+        let regs: &GpioRegisters = unsafe { &*self.regs };
+        regs.doe.set(regs.doe.get() | self.pin_mask);
+    }
+
+    fn make_input(&self) {
+        self.enable_gpio();
+        ioc::IOCFG[self.pin].enable_input();
+    }
+
+    fn disable(&self) {
+        hil::gpio::PinCtl::set_input_mode(self, hil::gpio::InputMode::PullNone);
+    }
+
+    fn set(&self) {
+        let regs: &GpioRegisters = unsafe { &*self.regs };
+        regs.dout_set.set(self.pin_mask);
+    }
+
+    fn clear(&self) {
+        let regs: &GpioRegisters = unsafe { &*self.regs };
+        regs.dout_clr.set(self.pin_mask);
+    }
+
+    fn toggle(&self) {
+        let regs: &GpioRegisters = unsafe { &*self.regs };
+        regs.dout_tgl.set(self.pin_mask);
+    }
+
+    fn read(&self) -> bool {
+        let regs: &GpioRegisters = unsafe { &*self.regs };
+        regs.din.get() & self.pin_mask != 0
+    }
+
+    fn enable_interrupt(&self, client_data: usize, mode: hil::gpio::InterruptMode) {
+        self.client_data.set(client_data);
+        ioc::IOCFG[self.pin].enable_interrupt(mode);
+    }
+
+    fn disable_interrupt(&self) {
+        ioc::IOCFG[self.pin].disable_interrupt();
+    }
+}
+
+pub struct Port {
+    pins: [GPIOPin; NUM_PINS],
+}
+
+impl Index<usize> for Port {
+    type Output = GPIOPin;
+
+    fn index(&self, index: usize) -> &GPIOPin {
+        &self.pins[index]
+    }
+}
+
+impl IndexMut<usize> for Port {
+    fn index_mut(&mut self, index: usize) -> &mut GPIOPin {
+        &mut self.pins[index]
+    }
+}
+
+impl Port {
+    pub fn handle_interrupt(&self) {
+        let regs: &GpioRegisters = unsafe { &*GPIO_BASE };
+        let evflags = regs.evflags.get();
+        // Clear all interrupts by setting their bits to 1 in evflags
+        regs.evflags.set(evflags);
+
+        // evflags indicate which pins has triggered an interrupt,
+        // we need to call the respective handler for positive bit in evflags.
+        let mut pin: usize = usize::max_value();
+        while pin < self.pins.len() {
+            pin = evflags.trailing_zeros() as usize;
+            if pin >= self.pins.len() {
+                break;
+            }
+
+            self.pins[pin].handle_interrupt();
```

Comment:
Manually calling the interrupt handlers should be avoided, it's assumed that these are triggered by hardware events!. Change or explain why this is necessary

Reply:
The interrupts are treated differently depending on what pin has been fired (i.e each pin has a different `hil::gpio::Client`). Basically whenever a gpio interrupt is generated, we arrive at `Port::handle_interrupt`, which then determines which pins that generated an interrupt (there can be more than one), and then call their respective interrupt handlers (`GpioPin::handle_interrupt`).

Whether this is the right practice or not I can't tell, it's done for the [nrf5x chip as well](https://github.com/helena-project/tock/blob/master/chips/nrf5x/src/gpio.rs#L199).

An alternative approach would be to move the logic of `GpioPin::handle_interrupt` into `Port::handle_interrupt`. Maybe adding a comment which explains would be good enough, or do you think it's better to combine the logic?

Reply:
I think that a comment about this is good enough

Reply:
Yeah, this is totally reasonable. I think we might do something similar on the SAM4L. Basically it's a two level dispatch: the hardware/event-loop can trigger the interrupt for the port, which then figures out which pin it was actually for and routes the interrupt.